React 17 context API & hook

JosephClay opened the pull request:
First steps of updating the library to remove legacy `getChildContext` API

Next steps:
- [x] move hoc branch `componentWillMount` logic to the constructor
- [x] replace `componentWillReceiveProps` with `componentDidUpdate`
- [x] create a `useBranch` hook
- [ ] ~~remove mixins~~

@Yomguithereal does this sound like a good plan?

JosephClay commented:
Refactoring to `componentDidUpdate` required a deep-equality check to keep from going into an infinite loop: https://www.npmjs.com/package/deep-equal

JosephClay commented:
Hook `useBranch` and hook tests have been added, mimicking higher-order's tests

Yomguithereal commented:
Hello @JosephClay 

> remove mixins?

Probably not. I still know people using them and I feel it would be sad to leave them behind (even by bumping a major version).

Did you write docs for the hooks?

@ arnaudmolo can I ask you for your opinion on those hooks?

JosephClay commented:
@Yomguithereal Haven't created documentation yet

Would this be the way you like the hook to be exposed?
```
import { useBranch } from 'baobab-react/higher-order';
```

Yomguithereal commented:
Maybe more:

```js
import { useBranch } from 'baobab-react/hooks';
```

Also should we move to new context APIs?

JosephClay commented:
New context APIs were the first things added to remove `getChildContext`

JosephClay commented:
@Yomguithereal @arnaudmolo I've added documentation for hooks, based off higher-order. In fact, with the refactor to using the new Context API,  higher-order and hooks are interoperable!

I've added the hook `useRoot` to be consistent

Added `dispatch` to the `useBranch` hook for parity with `higher-order/branch` and added action tests

Yomguithereal commented on docs/hooks.md:
```diff
@@ -0,0 +1,291 @@
+import List from './list.jsx';
+
+// Creating our top-level component
+const App = function({tree}) {
```
Could we avoid variable scope shadowing here, related to the `tree` name?

JosephClay replied:
We could pass it down as `store`?

Yomguithereal commented:
Seems good to me. I just have a remark on the docs I listed herebefore. @arnaudmolo everything's good for you?

arnaudmolo commented:
@Yomguithereal 👍

Yomguithereal commented:
I am going to merge this. @JosephClay should this be a v4 or can this be a safe v3.1?

JosephClay commented:
@Yomguithereal I believe a v4 would be in order, as the Context API requires React `^16.x`. Has the potential to break projects using `higher-order` with an older version of React

Yomguithereal commented:
Fair enough. Let's go then :)

Yomguithereal commented:
v4.0.0 is now live on npm. Thanks @JosephClay.
